### PR TITLE
[4.0][Features][Ready] Default operation configuration methods, different ways to split setup() by operation

### LINKED ISSUE #1375: [4.0][Performance Improvement] Columns, filters and fields should be loaded when needed
# Bug report
Not a bug, improvement. If you have several relations in relatively big table, page loading takes too much time.

### What I did:
Defult CRUD controller created.

### What I expected to happen:
addFilters, addFields and addColumns should be populated when needed.

### What happened:
On every request addFilters, addFields and addColumns are populated.

### What I've already tried to fix it:
```
        $currentScreen = $this->getCurrentScreen();
        if (in_array($currentScreen, ['index', 'show', 'search'])) {
            $this->addColumns();
            $this->addFilters();
        }
        elseif (in_array($currentScreen, ['create', 'edit'])) {
            $this->addFields();
        }
```

### Backpack, Laravel, PHP, DB version:
BP: 3.4.6., L:5.6.18, PHP: 7.2.4, Mysql: 5.7.x

### PR DESCRIPTION
**Pull request by tabacitu**
**TLDR: I've added a way to specify which operation you're referring to, for each ```$this->crud->smth()``` call - basically including those calls in a closure. I'm wondering if we should make this optional or mandatory. Skip to Feature 2 for details.**

----

This PR adds two features:

# (1) Operations can have defaults

**Premise/problem:** Some operations, when enabled, need to add things like settings, buttons, etc. For example, ```ReorderOperation``` needs a button to show up in the "top" stack if enabled. Previously this had been hard-coded in ```initButtons()``` - which was ugly and difficult to overwrite.

**Solution:** Similar to how we decided to do about routes, Operations can have a magic method inside them: ```setupOperationNameDefaults()```. If present, that method will be called before ```setup()```. 

**Example:** Once you use the ```ReorderOperation``` trait on your ```EntityCrudController```, this method will automatically be run, so the button will be added automatically.
```php
    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupReorderDefaults()
    {
        $this->crud->set('reorder.enabled', true);
        $this->crud->allowAccess('reorder');
        $this->crud->addButton('top', 'reorder', 'view', 'crud::buttons.reorder');
    }
```

Very happy with this feature.

# (2) Specify which Operation the current CrudPanel call is referring to

**Premise/problem:** Currently, you don't know which operation your CrudPanel calls affects, when calling them from the ```setup()``` method:
- when doing ```$this->crud->addField('text')```, it adds the field to both **Create** and **Update** operations; you can specify which operation using a second parameter;
- when doing ```$this->crud->addColumn('text')```, it adds the column to the **List** operation;
- when doing ```$this->crud->addButton()```, it adds the button to the **List** operation;
- when doing ```$this->crud->addFilter()```, it adds the filter to the **List** operation;
- when doing ```$this->crud->enableDetailsRow()```, it affects the **List** operation;

(and so on; for all the List operations above, you can't actually specify which operation you're referring to, not with a second parameter, not with anything)

I think this is confusing, because some CrudPanel features are actually used by multiple operations:
- both **Create** and **Update** use Fields, Tabs, Errors, etc;
- both **List** and **Show** use Columns and Buttons in Show;

Currently the **Show** operation does ```setFromDb()```, but it also includes any columns you specified with ```$this->crud->addColumn()``` in setup. Super-confusing, I think.

**Solution:** This PR adds an ```operation()``` method, that is currently optional, but **I'm thinking of making it mandatory. I could use a second opinion on that.** Basically I think all calls to fields, columns, buttons, filters, etc, should specify which operation they're referring to:

```php
// specify the operation using a closure
$this->crud->operation('list', function() {
    $this->crud->addColumn('text');
    $this->crud->enableDetailsRow();
});

// operation can also be an array of operations
$this->crud->operation(['create', 'update'], function() {
    $this->crud->addField('text');
});

```

**Example:**: Here's how a CategoryCrudController could look, when splitting calls in operations.
```php
<?php

namespace Backpack\NewsCRUD\app\Http\Controllers\Admin;

use Backpack\CRUD\app\Http\Controllers\CrudController;
use Backpack\CRUD\CrudPanelFacade as CRUD;

// VALIDATION: change the requests to match your own file names if you need form validation
use Backpack\NewsCRUD\app\Http\Requests\CategoryRequest as StoreRequest;
use Backpack\NewsCRUD\app\Http\Requests\CategoryRequest as UpdateRequest;

class CategoryCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\ReorderOperation;
    

    public function setup()
    {
        /*
        |--------------------------------------------------------------------------
        | BASIC CRUD INFORMATION
        |--------------------------------------------------------------------------
        */
        CRUD::setModel("Backpack\NewsCRUD\app\Models\Category");
        CRUD::setRoute(config('backpack.base.route_prefix', 'admin').'/category');
        CRUD::setEntityNameStrings('category', 'categories');

        /*
        |--------------------------------------------------------------------------
        | OPERATION CONFIGURATION
        |--------------------------------------------------------------------------
        */

        CRUD::operation('list', function() {
            CRUD::addColumn('name');
            CRUD::addColumn('slug');
            CRUD::addColumn([
                'label' => 'Parent',
                'type' => 'select',
                'name' => 'parent_id',
                'entity' => 'parent',
                'attribute' => 'name',
                'model' => "Backpack\NewsCRUD\app\Models\Category",
            ]);
        });

        CRUD::operation(['create', 'update'], function() {
            CRUD::addField([
                'name' => 'name',
                'label' => 'Name',
            ]);
            CRUD::addField([
                'name' => 'slug',
                'label' => 'Slug (URL)',
                'type' => 'text',
                'hint' => 'Will be automatically generated from your name, if left empty.',
                // 'disabled' => 'disabled'
            ]);
            CRUD::addField([
                'label' => 'Parent',
                'type' => 'select',
                'name' => 'parent_id',
                'entity' => 'parent',
                'attribute' => 'name',
                'model' => "Backpack\NewsCRUD\app\Models\Category",
            ]);
        });

        CRUD::operation('reorder', function() {
            CRUD::enableReorder('name', 2);
        });
    }

    public function store(StoreRequest $request)
    {
        return $this->storeEntry($request);
    }

    public function update(UpdateRequest $request)
    {
        return $this->updateEntry($request);
    }
}
```

Nice and clean, I think.

As a **bonus: whatever you specify inside that closure is only run during that operation**. This way,  all your addField() calls only happen when you're running the Create or Update operation. It won't be called during List or Reorder. This would fix #1375  and make the CrudController run faster, since at a certain time, not everything inside setup() will be called, but only:
- whatever is outside a closure
- whatever is inside the closure for the current operation

But I'm worried that:
- it would make the upgrade process slightly more difficult;
- it's a bit "magical"; is it more or less confusing than the current way? would you expect what is inside that closure to be run all the time, or just during that operation?

**Thoughts anyone?** @OwenMelbz , @OliverZiegler , @tswonke , @AbbyJanke , @lloy0076  maybe?

### CODE REVIEW

**Comment by tabacitu** [comment]
On second thought, do we really want to define operations in closures inside the ```setup()``` method? 

A simple alternative that would _also_ fix the problem above, would be to have yet another magic method on the operation (like ```setupReorderRoutes()``` and ```setupReorderDefaults()```), say... ```setupReorderConfiguration()```, that if present, will be called when that operation is run.
 
Then an EntityCrudController could look something like this:

```php
<?php

namespace Backpack\NewsCRUD\app\Http\Controllers\Admin;

use Backpack\CRUD\app\Http\Controllers\CrudController;
use Backpack\CRUD\CrudPanelFacade as CRUD;

// VALIDATION: change the requests to match your own file names if you need form validation
use Backpack\NewsCRUD\app\Http\Requests\CategoryRequest as StoreRequest;
use Backpack\NewsCRUD\app\Http\Requests\CategoryRequest as UpdateRequest;

class CategoryCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\ReorderOperation;
    
    /**
     * Configure the CrudPanel object.
     */
    public function setup()
    {
        CRUD::setModel("Backpack\NewsCRUD\app\Models\Category");
        CRUD::setRoute(config('backpack.base.route_prefix', 'admin').'/category');
        CRUD::setEntityNameStrings('category', 'categories');
    }

    /**
     * Configure the Reorder operation.
     */
    public function setupReorderConfiguration()
    {
        CRUD::set('reorder.label', 'name');
        CRUD::set('reorder.max_level', 2);
    }

    /**
     * Configure the List operation.
     */
    public function setupListConfiguration()
    {
        CRUD::addColumn('name');
        CRUD::addColumn('slug');
        CRUD::addColumn([
            'label' => 'Parent',
            'type' => 'select',
            'name' => 'parent_id',
            'entity' => 'parent',
            'attribute' => 'name',
            'model' => "Backpack\NewsCRUD\app\Models\Category",
        ]);
    }

    /**
     * Configure the Create operation.
     */
    public function setupCreateConfiguration()
    {
        CRUD::addField([
            'name' => 'name',
            'label' => 'Name',
        ]);
        CRUD::addField([
            'name' => 'slug',
            'label' => 'Slug (URL)',
            'type' => 'text',
            'hint' => 'Will be automatically generated from your name, if left empty.',
            // 'disabled' => 'disabled'
        ]);
        CRUD::addField([
            'label' => 'Parent',
            'type' => 'select',
            'name' => 'parent_id',
            'entity' => 'parent',
            'attribute' => 'name',
            'model' => "Backpack\NewsCRUD\app\Models\Category",
        ]);
    }

    /**
     * Configure the Update Operation.
     */
    public function setupUpdateConfiguration()
    {
        $this->setupCreateConfiguration();
    }

    public function store(StoreRequest $request)
    {
        return $this->storeEntry($request);
    }

    public function update(UpdateRequest $request)
    {
        return $this->updateEntry($request);
    }
}
```

Can't say that I like this more/less than using closures. It feels less magical, but it also feels a little bloated somehow, even if it's not...

Also, there's no reason we can't have all options:
- (A) everything spilled inside ```setup()```, like it's been in v3;
- (B) closures inside ```setup()```;
- (C) magic ```setupReorderConfiguration()``` methods;

In fact, this PR does have all of them. And the order for the configuration setup is:
- first everything that's in ```setup()```, but _outside_ closures;
- then everything that's in ```setupReorderDefaults()```;
- then, only during this operation, whatever's _inside_ ```setup()``` as closure;
- then, only during this operation, whatever's inside ```setupReorderConfiguration()```;

Right now I think:
- (A) I don't like it. It would provide backwards-compatibility; but it's the most confusing; and it would require each feature (Columns, Fields, Buttons, Filters, etc) to have a _default operation_, some scope it's being called on if no operation has been specified; I really don't like that;
- (B) I like it. Seems clean and laravelly; but confusing to have things both inside and outside closures; you wouldn't intuitively know this will only be run during that operation;
- (c) I like it. Seems like it would make controllers bigger, but actually it wouldn't - it would just split the ```setup()``` method into a bunch of other methods, one for each operation;

Anybody else have any thoughts about this?

**Comment by tabacitu** [comment]
I know nobody's going to read looong thread. But I've decided to:
- keep both (B) and (C);
- force developers to specify which operation is call refers too; I've provided a fallback so that old Controllers work (it assumes that it's about the current operation), which worked fine for 90% of my old Controllers, but I don't expect it to work for 100%;

Merged into ```cleanup-crudpanel-object``` here https://github.com/Laravel-Backpack/CRUD/pull/1997
